File: basic/change function declaration/javascript/example2.js
/*
Contexto: Estás desarrollando una aplicación de conversión de unidades de medida. Tienes una función llamada convertir_pies_a_metros que recibe una longitud en pies y realiza la conversión a metros. Actualmente, la función solo maneja la conversión de pies a metros. Aplica "Change Function Declaration" para permitir que la función convierta entre diferentes unidades de medida, como pies a metros, pulgadas a centímetros, etc.
*/

function convertir_pies_a_metros(pies) {
    // Realizar la conversión de pies a metros aquí
    metros = pies * 0.3048
    return metros
}

const unidadesDeMedida = {
    metros: {
        pulgadas: 39.37,
        pies: 3.281
    },

    pulgadas: {
        metros: 0.0254,
        pies: 0.0833333
    },

    pies: {
        metros: 0.3048,
        pulgadas: 12
    }
}

function conversionDeMedidas(cantidad, tipoEntrada, tipoSalida) {
    unidadMedidaSalida = unidadesDeMedida[tipoEntrada][tipoSalida] ?? null
    if (!unidadMedidaSalida) {
        throw Error("no existe la unidad de medida")
    }
    return cantidad * unidadMedidaSalida
}

module.exports = conversionDeMedidas

File: basic/change function declaration/python/src/example_2.py
"""
Ejercicio 2: Cálculo de Área de una Figura Geométrica

Contexto: Estás desarrollando una aplicación de geometría. Tienes una función llamada calcular_area que recibe la longitud y la altura de una figura geométrica y calcula su área. Actualmente, la función está diseñada solo para calcular el área de rectángulos. Aplica "Change Function Declaration" para permitir que la función calcule el área de diferentes tipos de figuras geométricas, como rectángulos, triángulos, círculos, etc.
"""
from abc import abstractmethod, ABC
def calcular_area_(longitud, altura):
    # Calcular el área de un rectángulo aquí
    area = longitud * altura
    return area

class GeometricFigure(ABC):
    def area(self) -> float:
        pass

class AbstractFactoryGeometricFigure(ABC):
    def create_square(self, *args) -> GeometricFigure:
        pass

    def create_rectangle(self, *args) -> GeometricFigure:
        pass

    def create_triangle(self, *args) -> GeometricFigure:
        pass

class Square(GeometricFigure):
    _side: float

    def __init__(self, side: float) -> None:
        self._side = side

    def area(self) -> float:
        return self._side * self._side

class Rectangle(GeometricFigure):
    _vertical_side: float
    _horizontal_side: float

    def __init__(self, vertical_side: float, horizontal_side: float) -> None:
        self._vertical_side = vertical_side
        self._horizontal_side = horizontal_side

    def area(self) -> float:
        return self._vertical_side * self._horizontal_side

class Triangle(GeometricFigure):
    _base: float
    _height: float

    def __init__(self, base: float, height: float) -> None:
        self._base = base
        self._height = height

    def area(self) -> float:
        return (self._base * self._height) / 2

class GeometricFigures(AbstractFactoryGeometricFigure):
    def create_square(self, *args) -> GeometricFigure:
        return Square(*args)

    def create_rectangle(self, *args) -> GeometricFigure:
        return Rectangle(*args)

    def create_triangle(self, *args) -> GeometricFigure:
        return Triangle(*args)

def calcular_area(figura: GeometricFigure) -> float:
    return figura.area()
